File: WordProbabilityCalculator.py
import collections
import re
import os
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')
nltk.download('stopwords')


def calculate_word_probabilities(text, stop_words):
    words = word_tokenize(text.lower())
    filtered_words = [word for word in words if word.isalnum()
                      and word not in stop_words]
    return filtered_words


stop_words = set(stopwords.words('english'))

folder_path = 'documents'

all_word_counts = collections.Counter()

for file_name in os.listdir(folder_path):
    if file_name.endswith('.txt'):
        file_path = os.path.join(folder_path, file_name)
        try:
            with open(file_path, 'r', encoding='utf-8') as file:
                document_text = file.read()
                filtered_words = calculate_word_probabilities(
                    document_text, stop_words)
                word_counts = collections.Counter(filtered_words)
                all_word_counts.update(word_counts)
        except UnicodeDecodeError:
            print(
                f"Error decoding {file_name} with UTF-8 encoding. Trying another encoding...")
            try:
                with open(file_path, 'r', encoding='latin-1') as file:
                    document_text = file.read()
                    filtered_words = calculate_word_probabilities(
                        document_text, stop_words)
                    word_counts = collections.Counter(filtered_words)
                    all_word_counts.update(word_counts)
            except Exception as e:
                print(f"Error decoding {file_name} with Latin-1 encoding: {e}")

total_words = sum(all_word_counts.values())
top_words = all_word_counts.most_common(5)
word_probabilities = [(word, count / total_words, count)
                      for word, count in top_words]

print("Top 5 words across all documents:")
for word, probability, count in word_probabilities:
    print(f'  Word: {word}, Probability: {probability:.4f}, Count: {count}')
print()

File: WordProbabilityCalculator.js
const fs = require("fs");
const path = require("path");

function loadStopWords(filePath) {
  try {
    const data = fs.readFileSync(filePath, "utf8");
    return new Set(
      data
        .trim()
        .toLowerCase()
        .split("\n")
        .map((word) => word.trim())
    );
  } catch (err) {
    console.error("Error loading stopwords file:", err.message);
    return new Set();
  }
}

function calculateWordProbabilities(text, stopWords) {
  const words = text.toLowerCase().split(/\s+/);
  return words.filter(
    (word) => word.match(/[A-Za-z0-9]+/) && !stopWords.has(word)
  );
}

function main() {
  const folderPath = "documents";
  const stopwordsFilePath = "stopwords.txt";

  try {
    const stopWords = loadStopWords(stopwordsFilePath);
    const allWordCounts = {};

    const files = fs.readdirSync(folderPath);

    files.forEach((fileName) => {
      if (fileName.endsWith(".txt")) {
        try {
          const filePath = path.join(folderPath, fileName);
          const documentText = fs.readFileSync(filePath, "utf8");
          const filteredWords = calculateWordProbabilities(
            documentText,
            stopWords
          );

          filteredWords.forEach((word) => {
            allWordCounts[word] = (allWordCounts[word] || 0) + 1;
          });
        } catch (err) {
          console.error("Error reading file:", err.message);
        }
      }
    });

    const totalWords = Object.values(allWordCounts).reduce(
      (sum, count) => sum + count,
      0
    );

    const sortedWordCounts = Object.keys(allWordCounts).sort(
      (a, b) => allWordCounts[b] - allWordCounts[a]
    );

    const topWords = sortedWordCounts.slice(
      0,
      Math.min(5, sortedWordCounts.length)
    );

    console.log("Top 5 words across all documents:");
    topWords.forEach((word) => {
      const count = allWordCounts[word];
      const probability = count / totalWords;
      console.log(
        `  Word: ${word}, Probability: ${probability.toFixed(
          4
        )}, Count: ${count}`
      );
    });
  } catch (err) {
    console.error("Error:", err.message);
  }
}

main();
